src/ray/core_worker/task_manager.h

```diff
@@ -100,6 +108,47 @@ class TaskManager : public TaskFinisherInterface {
   int NumPendingTasks() const { return pending_tasks_.size(); }
 
  private:
+  struct TaskEntry {
+    TaskEntry(const TaskSpecification &spec_arg, int num_retries_left_arg,
+              size_t num_returns)
+        : spec(spec_arg), num_retries_left(num_retries_left_arg) {
+      for (size_t i = 0; i < num_returns; i++) {
+        plasma_returns_in_scope.insert(spec.ReturnId(i, TaskTransportType::DIRECT));
+      }
+    }
+    // The task spec. This is pinned as long as the following are true:
+    // - num_retries_left > 0. If this is not true, then this means that the
+    //   task may be retried in the future.
+    // - num_executions > 0. If this is not true, then this means that the task
+    //   is still pending its first execution.
+    // - num_plasma_returns_in_scope > 0. If this is not true, then this means
+    //   that the task does not have any return values that may be passed as
+    //   dependencies to other tasks.
+    const TaskSpecification spec;
+    // Number of times this task may be resubmitted. If this reaches 0, then
+    // the task entry may be erased.
+    int num_retries_left;
+    // Whether this task is currently pending execution. This is used to pin
+    // the task entry if the task is still pending but all of its return IDs
+    // are out of scope.
+    bool pending = true;
+    // Plasma objects returned by this task whose value is still needed. This
+    // is set initially to the task's return objects. Once the task finishes
+    // its first execution, then the objects that the task returned by value
+    // are removed from this set. Objects that the task returned through plasma
+    // are removed from this set once:
+    // 1) The language frontend no longer has a reference to the object ID.
+    // 2) There are no tasks that depend on the object. This includes both
+    //    pending tasks and tasks that finished execution but that may be
+    //    retried in the future.
+    absl::flat_hash_set<ObjectID> plasma_returns_in_scope;
```

Comment:
Consider `potentially_needed_return_ids`.

Reply:
Changed to `reconstructable_return_ids`.